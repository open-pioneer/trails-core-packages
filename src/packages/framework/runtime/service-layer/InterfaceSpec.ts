import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import { InterfaceReferenceMetadata } from "../metadata";

/**
 * Describes the implementation of an interface.
 */
export interface InterfaceSpec {
    interfaceName: string;
    qualifier?: string | undefined;
}

export function renderInterfaceSpec({ interfaceName, qualifier }: InterfaceSpec): string {
    return `'${interfaceName}'` + (qualifier ? ` (qualifier: '${qualifier}')` : "");
}

export interface SingleImplementationSpec extends InterfaceSpec {
    all?: undefined;
}

export interface AllImplementationsSpec extends Pick<InterfaceSpec, "interfaceName"> {
    all: true;
}

/**
 * Describes a reference to one or more implementations of an interface.
 */
export type ReferenceSpec = SingleImplementationSpec | AllImplementationsSpec;

export function parseReferenceSpec(metadata: InterfaceReferenceMetadata): ReferenceSpec {
    if (!metadata) {
        throw new Error(ErrorId.INVALID_METADATA, "Missing reference metadata.");
    }
    if (metadata.all && metadata.qualifier) {
        throw new Error(
            ErrorId.INVALID_METADATA,
            "References can't use 'qualifier' and 'all' properties at the same time."
        );
    }
    if (metadata.all) {
        return {
            all: true,
            interfaceName: metadata.name
        };
    }
    return {
        interfaceName: metadata.name,
        qualifier: metadata.qualifier
    };
}

export function isSingleImplementationSpec(spec: ReferenceSpec): spec is SingleImplementationSpec {
    return !spec.all;
}

export function isAllImplementationsSpec(spec: ReferenceSpec): spec is AllImplementationsSpec {
    return spec.all === true;
}
