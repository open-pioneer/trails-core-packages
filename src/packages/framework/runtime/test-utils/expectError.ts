export function expectError(impl: () => unknown): Error {
    try {
        impl();
        throw new Error("expected error!");
    } catch (e) {
        if (e instanceof Error) {
            return e;
        }
        throw new Error("unexpected error value, not an instance of Error");
    }
}

export function expectAsyncError(impl: () => Promise<unknown>): Promise<Error> {
    const promise = impl();
    return promise.then(
        () => {
            throw new Error("expected error!");
        },
        (e) => {
            if (e instanceof Error) {
                return e;
            }
            throw new Error("unexpected error value, not an instance of Error");
        }
    );
}
