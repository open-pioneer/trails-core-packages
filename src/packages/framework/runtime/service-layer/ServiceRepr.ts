import { ErrorId } from "../errors";
import { ServiceMetadata } from "../metadata";
import { Service, ServiceConstructor, ServiceOptions } from "../Service";
import { Error } from "@open-pioneer/core";
import { InterfaceSpec, parseReferenceSpec, ReferenceSpec } from "./InterfaceSpec";
import { PackageIntl } from "../i18n";

export type ServiceState = "not-constructed" | "constructing" | "constructed" | "destroyed";

export type ServiceDependency = ReferenceSpec & { referenceName: string };

export interface ConstructorFactory {
    type: "class";
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    clazz: ServiceConstructor<any>;
}

export interface FunctionFactory {
    type: "function";
    create: (options: ServiceOptions) => Service;
}

export type ServiceFactory = ConstructorFactory | FunctionFactory;

export interface ServiceReprOptions {
    name: string;
    packageName: string;
    factory: ServiceFactory;
    intl: PackageIntl;
    dependencies?: ServiceDependency[];
    interfaces?: InterfaceSpec[];
    properties?: Record<string, unknown>;
}

/**
 * Represents metadata and state of a service in the runtime.
 * `this.instance` is the actual service instance (when constructed).
 */
export class ServiceRepr {
    static create(
        packageName: string,
        data: ServiceMetadata,
        intl: PackageIntl,
        properties?: Record<string, unknown>
    ): ServiceRepr {
        const clazz = data.clazz;
        const name = data.name;
        const dependencies = Object.entries(data.references ?? {}).map(
            ([name, referenceMetadata]) => {
                return {
                    referenceName: name,
                    ...parseReferenceSpec(referenceMetadata)
                };
            }
        );
        const interfaces = data.provides?.map<InterfaceSpec>((i) => {
            return {
                interfaceName: i.name,
                qualifier: i.qualifier
            };
        });
        const factory: ConstructorFactory = {
            type: "class",
            clazz
        };
        return new ServiceRepr({
            name,
            packageName,
            factory,
            intl,
            dependencies,
            interfaces,
            properties
        });
    }

    /** Unique id of this service. Contains the package name and the service name. */
    readonly id: string;

    /** Name of this service in it's package. */
    readonly name: string;

    /** Name of the parent package. */
    readonly packageName: string;

    /** Locale-dependant i18n messages. */
    readonly intl: PackageIntl;

    /** Service properties made available via the service's constructor. */
    readonly properties: Readonly<Record<string, unknown>>;

    /** Dependencies required by the service constructor. */
    readonly dependencies: readonly ServiceDependency[];

    /** Interfaces provided by the service. */
    readonly interfaces: readonly Readonly<InterfaceSpec>[];

    /** Number of references to this service. */
    private _useCount = 0;

    /** Service factory to construct an instance. */
    private factory: ServiceFactory;

    /** Current state of this service. "constructed" -> instance is available. */
    private _state: ServiceState = "not-constructed";

    /** Service instance, once constructed. */
    private _instance: Service | undefined = undefined;

    constructor(options: ServiceReprOptions) {
        const {
            name,
            packageName,
            factory,
            intl,
            dependencies = [],
            interfaces = [],
            properties = {}
        } = options;
        if (!isValidServiceName(name)) {
            throw new Error(ErrorId.INTERNAL, `Invalid service name: '${name}'.`);
        }

        this.id = `${packageName}::${name}`;
        this.name = name;
        this.packageName = packageName;
        this.factory = factory;
        this.intl = intl;
        this.dependencies = dependencies;
        this.interfaces = interfaces;
        this.properties = properties;
    }

    /** Returns the current service instance or undefined if the service has not been constructed. */
    get instance(): Service | undefined {
        return this._instance;
    }

    /** Returns the current state of the service. */
    get state() {
        return this._state;
    }

    /** Returns the current reference count. */
    get useCount() {
        return this._useCount;
    }

    /**
     * Same as `instance`, but throws when the instance has not been constructed.
     */
    getInstanceOrThrow() {
        const instance = this._instance;
        if (!instance) {
            throw new Error(ErrorId.INTERNAL, "Expected service instance to be present.");
        }
        return instance;
    }

    /**
     * Called before `create()` to place the service into the `constructing` state,
     * which is currently used to detect cycles.
     */
    beforeCreate() {
        if (this._state === "not-constructed") {
            this._state = "constructing";
        } else {
            throw new Error(
                ErrorId.INTERNAL,
                "Inconsistent state: service is already under construction."
            );
        }
    }

    /**
     * Instantiates the service by invoking the service constructor
     * with the given `options`.
     *
     * The service's use count is initialized to `1`, so every `create()`
     * should be paired with a `removeRef()`.
     *
     * `destroy()` can be invoked once the final `removeRef()` has returned zero.
     */
    create(options: Pick<ServiceOptions, "references" | "referencesMeta">) {
        if (this._state !== "constructing" || this.instance !== undefined) {
            throw new Error(
                ErrorId.INTERNAL,
                "Inconsistent state: service is not being constructed."
            );
        }
        try {
            this._instance = createService(this.factory, {
                ...options,
                properties: this.properties,
                intl: this.intl
            });
            this._state = "constructed";
            this._useCount = 1;
            return this._instance;
        } catch (e) {
            throw new Error(
                ErrorId.SERVICE_CONSTRUCTION_FAILED,
                `Failed to create service instance '${this.id}'.`,
                { cause: e }
            );
        }
    }

    destroy() {
        if (this._instance) {
            try {
                this._instance.destroy?.();
            } catch (e) {
                throw new Error(
                    ErrorId.SERVICE_DESTRUCTION_FAILED,
                    `Failed to destroy service instance '${this.id}'.`,
                    { cause: e }
                );
            }
        }
        this._instance = undefined;
        this._state = "destroyed";
    }

    /**
     * Adds a use to the service's use count.
     * References to a service are tracked: it should only be destroyed when it is no longer being used.
     */
    addRef() {
        return (this._useCount += 1);
    }

    /**
     * Removes a use from the service's use count.
     * Returns the new use count.
     */
    removeRef() {
        return (this._useCount -= 1);
    }
}

export function createConstructorFactory<T extends {}>(
    clazz: ServiceConstructor<T>
): ConstructorFactory {
    return { type: "class", clazz };
}

export function createFunctionFactory(
    create: (options: ServiceOptions) => Service
): FunctionFactory {
    return { type: "function", create };
}

const SERVICE_NAME_REGEX = /^[a-z0-9_-]+$/i;

function isValidServiceName(name: string) {
    return SERVICE_NAME_REGEX.test(name);
}

function createService(factory: ServiceFactory, options: ServiceOptions) {
    switch (factory.type) {
        case "class":
            return new factory.clazz(options);
        case "function":
            return factory.create(options);
    }
}
