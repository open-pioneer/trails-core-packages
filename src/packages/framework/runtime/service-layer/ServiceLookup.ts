// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import { InterfaceSpec, isSingleImplementationSpec, ReferenceSpec } from "./InterfaceSpec";
import { ServiceRepr } from "./ServiceRepr";

interface Services {
    // The unqualified service implementations (if any).
    unqualified: ServiceRepr[];

    // Services indexed by their qualifier (if any).
    byQualifier: Map<string, ServiceRepr>;
}

export type ReadonlyServiceLookup = Pick<
    ServiceLookup,
    "lookup" | "lookupOne" | "lookupAll" | "serviceCount"
>;

export interface Unimplemented {
    type: "unimplemented";
}

export interface Ambiguous {
    type: "ambiguous";
    choices: AmbiguousChoice[];
}

export type AmbiguousChoice = [serviceId: string, qualifier: string | undefined];

export interface Found<T> {
    type: "found";
    value: T;
}

export type ServiceLookupResult = Unimplemented | Ambiguous | Found<ServiceRepr>;

export type ServicesLookupResult = Found<ServiceRepr[]>;

export class ServiceLookup {
    // Service implementations indexed by interface name.
    private services = new Map<string, Services>();

    // Total service count
    private _count = 0;

    /**
     * Returns the total number of registered services.
     */
    get serviceCount() {
        return this._count;
    }

    /**
     * Attempts to register the given service as an implementation of `interfaceName`.
     * Throws if the registration did not complete successfully.
     *
     * All services should be registered before any lookups are attempted.
     */
    register(service: ServiceRepr, { interfaceName, qualifier }: InterfaceSpec): void {
        if (!interfaceName) {
            throw new Error(
                ErrorId.INVALID_METADATA,
                `Service '${service.id}' provides invalid interface '${interfaceName}'.`
            );
        }
        if (qualifier != null && !qualifier) {
            throw new Error(
                ErrorId.INVALID_METADATA,
                `Service '${service.id}' uses an invalid qualifier for interface '${interfaceName}': '${qualifier}'.`
            );
        }

        const services = this.ensureInterfaceEntry(interfaceName);
        if (!qualifier) {
            services.unqualified.push(service);
            ++this._count;
            return;
        }

        const existingQualified = services.byQualifier.get(qualifier);
        if (existingQualified) {
            throw new Error(
                ErrorId.DUPLICATE_INTERFACE,
                `Cannot register '${service.id}' as interface '${interfaceName}' with qualifier '${qualifier}'.` +
                    ` The interface is already provided by service '${existingQualified.id}'.` +
                    ` You can choose a different qualifier or remove the service to resolve the collision.`
            );
        }
        services.byQualifier.set(qualifier, service);
        ++this._count;
    }

    /**
     * Attempts to find the service(s) required by the given reference spec.
     */
    lookup(ref: ReferenceSpec): ServiceLookupResult | ServicesLookupResult {
        if (isSingleImplementationSpec(ref)) {
            return this.lookupOne(ref);
        } else {
            return this.lookupAll(ref.interfaceName);
        }
    }

    /**
     * Attempts to find a service implementing the given interface.
     */
    lookupOne({ interfaceName, qualifier }: InterfaceSpec): ServiceLookupResult {
        if (!interfaceName) {
            throw new Error(
                ErrorId.INVALID_METADATA,
                `Invalid interface name during service lookup: '${interfaceName}'.`
            );
        }
        if (qualifier != null && !qualifier) {
            throw new Error(
                ErrorId.INVALID_METADATA,
                `Invalid qualifier during service lookup of interface '${interfaceName}': '${qualifier}'.`
            );
        }

        const services = this.services.get(interfaceName);
        if (!services) {
            return { type: "unimplemented" };
        }

        if (!qualifier) {
            // Just pick one, but the choice must be unambiguous.
            const count = services.unqualified.length + services.byQualifier.size;
            if (count > 1) {
                // [serviceId, qualifier] tuples
                const choices: AmbiguousChoice[] = [];
                choices.push(
                    ...Array.from(services.byQualifier.entries()).map<AmbiguousChoice>(
                        ([qualifier, service]) => [service.id, qualifier]
                    )
                );
                choices.push(
                    ...services.unqualified.map<AmbiguousChoice>((s) => [s.id, undefined])
                );
                return {
                    type: "ambiguous",
                    choices: choices
                };
            }

            const service = services.unqualified[0] ?? first(services.byQualifier);
            if (!service) {
                return { type: "unimplemented" };
            }
            return { type: "found", value: service };
        }

        const service = services.byQualifier.get(qualifier);
        if (!service) {
            return { type: "unimplemented" };
        }
        return { type: "found", value: service };
    }

    /**
     * Returns all services implementing the given interface.
     */
    lookupAll(interfaceName: string): ServicesLookupResult {
        if (!interfaceName) {
            throw new Error(
                ErrorId.INVALID_METADATA,
                `Invalid interface name during service lookup: '${interfaceName}'.`
            );
        }

        const services = this.services.get(interfaceName);
        if (!services) {
            return { type: "found", value: [] };
        }

        const deduped = new Set([...services.unqualified, ...services.byQualifier.values()]);
        return {
            type: "found",
            value: Array.from(deduped)
        };
    }

    private ensureInterfaceEntry(interfaceName: string): Services {
        const services = this.services;
        let entry = services.get(interfaceName);
        if (!entry) {
            entry = {
                unqualified: [],
                byQualifier: new Map()
            };
            services.set(interfaceName, entry);
        }
        return entry;
    }
}

export function renderAmbiguousServiceChoices(choices: AmbiguousChoice[], max = 3): string {
    let message = "";
    let count = 0;
    for (const [serviceId, qualifier] of choices) {
        if (count) {
            message += ", ";
        }

        message += `'${serviceId}'`;
        if (qualifier) {
            message += ` (with qualifier '${qualifier}')`;
        }

        if (++count >= max) {
            break;
        }
    }

    const remaining = choices.length - count;
    if (remaining > 0) {
        message += ` and ${remaining} more`;
    }
    return message;
}

function first<K, V>(map: Map<K, V>): V | undefined {
    return map.values().next().value;
}
