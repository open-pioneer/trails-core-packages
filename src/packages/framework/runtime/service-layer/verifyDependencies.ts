import { Error } from "@open-pioneer/core";
import { ErrorId } from "../errors";
import { InterfaceSpec, ReferenceSpec, renderInterfaceSpec } from "./InterfaceSpec";
import {
    AmbiguousChoice,
    ReadonlyServiceLookup,
    renderAmbiguousServiceChoices,
    ServiceLookup
} from "./ServiceLookup";
import { ServiceRepr } from "./ServiceRepr";

export interface DependencyOptions {
    /**
     * All services known to the service layer.
     */
    services?: readonly ServiceRepr[];

    /**
     * References that must be satisfied by the services.
     */
    requiredReferences?: readonly RequiredReference[];
}

export type RequiredReference = UIDependency | FrameworkDependency;

export type UIDependency = { type: "ui"; packageName: string } & ReferenceSpec;

export type FrameworkDependency = { type: "framework" } & ReferenceSpec;

export type ComputedServiceDependencies = Record<string, ServiceRepr | ServiceRepr[]>;

export interface VerifyDependenciesResult {
    /**
     * Precomputed dependencies for every service.
     *
     * Key: service id.
     */
    serviceDependencies: Map<string, ComputedServiceDependencies>;

    /**
     * Index structure mapping interface names to service instances.
     */
    serviceLookup: ReadonlyServiceLookup;
}

/**
 * Visits all services and ensures that their dependencies can be satisfied by each other without a cycle.
 *
 * Throws an error if a cycle is detected or if a required service is never implemented.
 *
 * @returns an object containing the service index and precomputed dependency information
 */
export function verifyDependencies(options: DependencyOptions): VerifyDependenciesResult {
    const verifier = new Verifier(options);
    verifier.verify();
    return {
        serviceLookup: verifier.getServiceLookup(),
        serviceDependencies: verifier.getComputedDependencies()
    };
}

interface GraphItem {
    state: ItemState;
    service: ServiceRepr;
    dependencies: ComputedServiceDependencies;
}

type ItemState = "not-visited" | "pending" | "done";

/** The reason why a graph node is being visited. */
type ItemVisitReason = "root" | InterfaceReference;

type InterfaceReference =
    | {
          type: "framework-reference";
          value: ReferenceSpec;
      }
    | {
          type: "ui-reference";
          packageName: string;
          value: ReferenceSpec;
      }
    | {
          type: "service-reference";
          service: ServiceRepr;
          referenceName: string;
          value: ReferenceSpec;
      };

class Verifier {
    private requiredReferences: readonly RequiredReference[];
    private items: GraphItem[];
    private serviceLookup = new ServiceLookup();
    private serviceToGraphItem = new Map<ServiceRepr, GraphItem>();

    // Stack of visited nodes
    private stack: [item: GraphItem, reason: ItemVisitReason][] = [];

    constructor(options: DependencyOptions) {
        this.requiredReferences = options.requiredReferences ?? [];
        const services = options.services ?? [];
        const items = (this.items = services.map<GraphItem>((service) => {
            return {
                service: service,
                state: "not-visited",
                dependencies: {}
            };
        }));

        for (const item of items) {
            for (const spec of item.service.interfaces) {
                this.registerService(item, spec);
            }
        }
    }

    verify() {
        for (const ref of this.requiredReferences) {
            let interfaceRef: InterfaceReference;
            switch (ref.type) {
                case "framework":
                    interfaceRef = { type: "framework-reference", value: ref };
                    break;
                case "ui":
                    interfaceRef = {
                        type: "ui-reference",
                        packageName: ref.packageName,
                        value: ref
                    };
                    break;
            }
            this.visitReference(interfaceRef);
        }
        for (const item of this.items) {
            this.visitItem(item, "root");
        }
    }

    getServiceLookup() {
        return this.serviceLookup;
    }

    getComputedDependencies() {
        return new Map(
            this.items.map<[string, ComputedServiceDependencies]>((item) => {
                return [item.service.id, item.dependencies];
            })
        );
    }

    private visitItem(item: GraphItem, reason: ItemVisitReason) {
        const stack = this.stack;
        const state = item.state;
        if (state === "done") {
            return; // Item is initialized, cycle impossible
        }
        if (state === "pending") {
            this.throwCycleError(item, reason);
        }

        stack.push([item, reason]);
        item.state = "pending";
        for (const dep of item.service.dependencies) {
            item.dependencies[dep.referenceName] = this.visitReference({
                type: "service-reference",
                service: item.service,
                referenceName: dep.referenceName,
                value: dep
            });
        }
        item.state = "done";
        stack.pop();
    }

    private visitReference(ref: InterfaceReference) {
        const itemOrItems = this.findServices(ref);
        if (Array.isArray(itemOrItems)) {
            for (const child of itemOrItems) {
                this.visitItem(child, ref);
            }
            return itemOrItems.map((item) => item.service);
        } else {
            this.visitItem(itemOrItems, ref);
            return itemOrItems.service;
        }
    }

    private registerService(item: GraphItem, spec: InterfaceSpec) {
        this.serviceLookup.register(item.service, spec);
        this.serviceToGraphItem.set(item.service, item);
    }

    private findServices(ref: InterfaceReference): GraphItem | GraphItem[] {
        const spec = ref.value;
        const lookupResult = this.serviceLookup.lookup(spec);
        switch (lookupResult.type) {
            case "unimplemented": {
                const message = unimplementedInterfaceMessage(ref);
                throw new Error(ErrorId.INTERFACE_NOT_FOUND, message);
            }
            case "ambiguous": {
                const message = ambiguousInterfaceMessage(ref, lookupResult.choices);
                throw new Error(ErrorId.AMBIGUOUS_DEPENDENCY, message);
            }
        }

        const value = lookupResult.value;
        if (Array.isArray(value)) {
            return value.map((service) => this.getGraphItem(service));
        } else {
            return this.getGraphItem(value);
        }
    }

    private getGraphItem(service: ServiceRepr): GraphItem {
        const graphItem = this.serviceToGraphItem.get(service);
        if (!graphItem || graphItem.service !== service) {
            throw new Error(
                ErrorId.INTERNAL,
                `Failed to find matching graph item for service '${service.id}'.`
            );
        }
        return graphItem;
    }

    private throwCycleError(item: GraphItem, reason: ItemVisitReason) {
        // Find the item on the stack (it must be there because we detected a cycle).
        const stack = this.stack;
        const ownIndex = stack.findIndex((entry) => entry[0] === item);
        if (ownIndex === -1) {
            throw new Error(ErrorId.INTERNAL, "Failed to find cycle participant on the stack.");
        }

        const cycleInfo = stack
            .slice(ownIndex)
            .concat([[item, reason]])
            .map<string>((entry, index) => {
                const item = entry[0];
                const reason = index === 0 ? "root" : entry[1]; // info before the cycle does not matter

                let message = `'${item.service.id}'`;
                if (typeof reason === "object") {
                    message += ` (${referenceDetail(reason)})`;
                }
                return message;
            });
        throw new Error(
            ErrorId.DEPENDENCY_CYCLE,
            `Detected dependency cycle: ${cycleInfo.join(" => ")}.`
        );
    }
}

function unimplementedInterfaceMessage(ref: InterfaceReference): string {
    const interfaceText = renderInterfaceSpec(ref.value);
    switch (ref.type) {
        case "service-reference":
            return `Service '${ref.service.id}' requires an unimplemented interface ${interfaceText} as dependency '${ref.referenceName}'.`;
        case "ui-reference":
            return `The UI of package '${ref.packageName}' requires an unimplemented interface ${interfaceText}.`;
        case "framework-reference":
            return `The framework requires an unimplemented interface ${interfaceText}.`;
    }
}

function ambiguousInterfaceMessage(ref: InterfaceReference, choices: AmbiguousChoice[]): string {
    const interfaceText = renderInterfaceSpec(ref.value);
    const serviceChoices = renderAmbiguousServiceChoices(choices);

    let message = "";
    switch (ref.type) {
        case "service-reference":
            message = `Service '${ref.service.id}' requires the ambiguous interface ${interfaceText} as dependency '${ref.referenceName}'.`;
            break;
        case "ui-reference":
            message = `The UI of package '${ref.packageName}' requires the ambiguous interface ${interfaceText}.`;
            break;
        case "framework-reference":
            message = `The framework requires the ambiguous interface ${interfaceText}.`;
            break;
    }
    message += ` Possible choices are ${serviceChoices}.`;
    return message;
}

function referenceDetail(reason: InterfaceReference) {
    switch (reason.type) {
        case "service-reference":
            return `'${reason.referenceName}' providing '${reason.value.interfaceName}'`;
        case "ui-reference":
            return `UI of package '${reason.packageName}' requiring '${reason.value.interfaceName}'`;
        case "framework-reference":
            return `framework requiring '${reason.value.interfaceName}'`;
    }
}
