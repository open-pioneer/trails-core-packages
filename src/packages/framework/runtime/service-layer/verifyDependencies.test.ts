import { assert, expect, it } from "vitest";
import { createEmptyI18n } from "../i18n";
import { expectError } from "../test-utils/expectError";
import { InterfaceSpec, ReferenceSpec } from "./InterfaceSpec";
import { ReadonlyServiceLookup } from "./ServiceLookup";
import { createConstructorFactory, ServiceDependency, ServiceRepr } from "./ServiceRepr";
import { verifyDependencies } from "./verifyDependencies";

it("does not return an error on acyclic graphs", function () {
    const services = mockServices([
        {
            name: "Map",
            package: "map",
            provides: ["services.Map"],
            requires: []
        },
        {
            name: "ExampleTool",
            package: "tools",
            provides: [],
            requires: ["services.Map"]
        }
    ]);

    const { serviceLookup, serviceDependencies: computedDependencies } = verifyDependencies({
        services: services
    });
    assert.strictEqual(serviceLookup.serviceCount, 1);

    const service = getService(serviceLookup, "services.Map");
    assert.strictEqual(service.id, "map::Map");

    const mapDependencies = computedDependencies.get("map::Map");
    assert.deepEqual(mapDependencies, {});

    const exampleToolDependencies = computedDependencies.get("tools::ExampleTool")!;
    assert.deepEqual(exampleToolDependencies, {
        dep_0: service
    });
});

it("throws when a service is not implemented", function () {
    const services = mockServices([
        {
            name: "ExampleTool",
            package: "tools",
            provides: [],
            requires: ["services.Map"]
        }
    ]);

    const message = expectError(() =>
        verifyDependencies({
            services: services,
            requiredReferences: []
        })
    ).message;
    expect(message).toMatchSnapshot();
});

it("does not throw when an interface is implemented multiple times", function () {
    const services = mockServices([
        {
            name: "Map1",
            package: "map",
            provides: ["services.Map"],
            requires: []
        },
        {
            name: "Map2",
            package: "map",
            provides: ["services.Map"],
            requires: []
        }
    ]);

    const { serviceLookup } = verifyDependencies({
        services: services,
        requiredReferences: []
    });
    expect(serviceLookup.serviceCount).toEqual(2);

    const mapServices = serviceLookup.lookupAll("services.Map").value.map((s) => s.id);
    mapServices.sort();
    expect(mapServices).toEqual(["map::Map1", "map::Map2"]);
});

it("allows multiple implementations if the services use a 'qualifier' for disambiguation", function () {
    const services = mockServices([
        {
            name: "Map1",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map1"
                }
            ]
        },
        {
            name: "Map2",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map2"
                }
            ]
        }
    ]);

    const { serviceLookup } = verifyDependencies({
        services: services,
        requiredReferences: []
    });
    assert.strictEqual(serviceLookup.serviceCount, 2);

    const map1 = getService(serviceLookup, "services.Map", "map1");
    assert.strictEqual(map1.id, "map::Map1");

    const map2 = getService(serviceLookup, "services.Map", "map2");
    assert.strictEqual(map2.id, "map::Map2");
});

it("throws for ambiguous service reference", function () {
    const services = mockServices([
        {
            name: "Map1",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map1"
                }
            ]
        },
        {
            name: "Map2",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map2"
                }
            ]
        },
        {
            name: "Map3",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map"
                }
            ]
        },
        {
            name: "Map4",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map"
                }
            ]
        },
        {
            name: "MapUser",
            package: "map-user",
            requires: [
                {
                    interfaceName: "services.Map"
                }
            ]
        }
    ]);

    const message = expectError(() =>
        verifyDependencies({
            services: services,
            requiredReferences: []
        })
    ).message;
    expect(message).toMatchSnapshot();
});

it("allows to pick an unambiguous implementation via classifier", function () {
    const services = mockServices([
        {
            name: "Map1",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map1"
                }
            ]
        },
        {
            name: "Map2",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map2"
                }
            ]
        },
        {
            name: "MapUser",
            package: "map-user",
            requires: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map2"
                }
            ]
        }
    ]);

    const { serviceLookup, serviceDependencies: computedDependencies } = verifyDependencies({
        services: services,
        requiredReferences: []
    });

    const map2 = getService(serviceLookup, "services.Map", "map2");

    const mapUserDependencies = computedDependencies.get("map-user::MapUser");
    assert.deepEqual(mapUserDependencies, {
        dep_0: map2
    });
});

it("allows to pick all implementations", function () {
    const services = mockServices([
        {
            name: "Map1",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map1"
                }
            ]
        },
        {
            name: "Map2",
            package: "map",
            provides: [
                {
                    interfaceName: "services.Map",
                    qualifier: "map2"
                }
            ]
        },
        {
            name: "MapUser",
            package: "map-user",
            requires: [
                {
                    interfaceName: "services.Map",
                    all: true
                }
            ]
        }
    ]);

    const { serviceDependencies, serviceLookup } = verifyDependencies({
        services: services,
        requiredReferences: []
    });

    const all = serviceLookup.lookupAll("services.Map").value;
    assert.sameMembers(
        all.map((s) => s.id),
        ["map::Map1", "map::Map2"]
    );

    const mapUserDeps = serviceDependencies.get("map-user::MapUser")!.dep_0!;
    assert.isArray(mapUserDeps);
    assert.sameMembers(mapUserDeps as unknown[], all);
});

it("throws when a component directly depends on itself", function () {
    const services = mockServices([
        {
            name: "Map",
            package: "map",
            provides: ["services.Map"],
            requires: ["services.Map"]
        }
    ]);
    const message = expectError(() =>
        verifyDependencies({
            services: services
        })
    ).message;
    expect(message).toMatchSnapshot();
});

it("throws when a component depends on itself via a larger cycle", function () {
    const services = mockServices([
        {
            name: "a",
            package: "A",
            provides: ["A"],
            requires: ["B"]
        },
        {
            name: "b",
            package: "B",
            provides: ["B"],
            requires: ["D", "C"]
        },
        {
            name: "c",
            package: "C",
            provides: ["C"],
            requires: ["D", "D", "A"]
        },
        {
            name: "d",
            package: "D",
            provides: ["D"]
        }
    ]);

    const message = expectError(() =>
        verifyDependencies({
            services: services
        })
    ).message;
    expect(message).toMatchSnapshot();
});

it("does not return an error when the UI requires an existing interface", function () {
    const services = mockServices([
        {
            name: "Map",
            package: "map",
            provides: ["services.Map"]
        }
    ]);

    const { serviceLookup } = verifyDependencies({
        services: services,
        requiredReferences: [
            {
                type: "ui",
                interfaceName: "services.Map",
                packageName: "foo"
            }
        ]
    });
    assert.strictEqual(serviceLookup.serviceCount, 1);

    const service = getService(serviceLookup, "services.Map");
    assert.strictEqual(service.id, "map::Map");
});

it("throws when the ui requires an interface that is not implemented", function () {
    const services = mockServices([
        {
            name: "Map",
            package: "map",
            provides: ["services.Map"],
            requires: []
        }
    ]);

    const message = expectError(() =>
        verifyDependencies({
            services,
            requiredReferences: [
                {
                    type: "ui",
                    packageName: "foo",
                    interfaceName: "services.Map2"
                }
            ]
        })
    ).message;
    expect(message).toMatchSnapshot();
});

interface ServiceData {
    package: string;
    name: string;
    requires?: (string | ReferenceSpec)[];
    provides?: (string | InterfaceSpec)[];
}

function mockServices(data: ServiceData[]): ServiceRepr[] {
    return data.map<ServiceRepr>((service) => {
        const name = service.name;
        const packageName = service.package;
        const clazz = class MockService {};
        const dependencies = service.requires?.map<ServiceDependency>((spec, index) => {
            return {
                referenceName: `dep_${index}`,
                ...toInterfaceSpec(spec)
            };
        });
        const interfaces = service.provides?.map(toInterfaceSpec);
        return new ServiceRepr({
            name,
            packageName,
            factory: createConstructorFactory(clazz),
            intl: createEmptyI18n(),
            dependencies,
            interfaces
        });
    });
}

function toInterfaceSpec(spec: string | InterfaceSpec): InterfaceSpec {
    if (typeof spec === "string") {
        return { interfaceName: spec };
    }
    return spec;
}

function getService(
    serviceLookup: ReadonlyServiceLookup,
    interfaceName: string,
    qualifier?: string | undefined
) {
    const result = serviceLookup.lookupOne({ interfaceName, qualifier });
    if (result.type !== "found") {
        throw new Error(`Failed to find interface ${interfaceName} (qualifier: ${qualifier}).`);
    }
    return result.value;
}
