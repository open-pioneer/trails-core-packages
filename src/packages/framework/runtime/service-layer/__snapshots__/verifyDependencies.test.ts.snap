// Vitest Snapshot v1

exports[`throws for ambiguous service reference 1`] = `"runtime:ambiguous-dependency: Service 'map-user::MapUser' requires the ambiguous interface 'services.Map' as dependency 'dep_0'. Possible choices are 'map::Map1' (with qualifier 'map1'), 'map::Map2' (with qualifier 'map2'), 'map::Map3' and 1 more."`;

exports[`throws when a component depends on itself via a larger cycle 1`] = `"runtime:dependency-cycle: Detected dependency cycle: 'A::a' => 'B::b' ('dep_0' providing 'B') => 'C::c' ('dep_1' providing 'C') => 'A::a' ('dep_2' providing 'A')."`;

exports[`throws when a component directly depends on itself 1`] = `"runtime:dependency-cycle: Detected dependency cycle: 'map::Map' => 'map::Map' ('dep_0' providing 'services.Map')."`;

exports[`throws when a service is not implemented 1`] = `"runtime:interface-not-found: Service 'tools::ExampleTool' requires an unimplemented interface 'services.Map' as dependency 'dep_0'."`;

exports[`throws when the ui requires an interface that is not implemented 1`] = `"runtime:interface-not-found: The UI of package 'foo' requires an unimplemented interface 'services.Map2'."`;
